### PR TITLE
Windows unicode

### PR DESCRIPTION
**Pull request by flagarde**
`Windows` use UNICODE (`wchar_t` in `utf16`), this PR add a better narrow<->wide converter only for Windows as utf16 is useless on other OS. I added special `WindowsError` exception too to be able to get the string corresponding to `GetLastError()` (converted to utf8 of course). This will allow to have better exceptions on windows without too much trouble. I expect to have one for `errno` too

### CODE REVIEW

**Comment by flagarde** [comment]
The goal is to move other utf8 utf32 function in unicode.* and suppress conversion.*
Need better exception handling and test before moving to unicode.*

File: cpp-terminal/platforms/exception.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+// Helper for windows errors
+#if defined(_WIN32)
+class WindowsError : public Term::Exception
+{
+public:
+  WindowsError(const unsigned long& error);
+  virtual ~WindowsError() = default;
+};
+#endif
```
Comment (certik) [inline cpp-terminal/platforms/exception.hpp:26]: Good practice in C++ is to keep all platform specific things in one "cpp" file. And the rest of the code to be fully platform independent, no ifdefs, and no ifdefs/macros in header files.

That's how cpp-terminal was originally written as well.

Is there a way to design this PR in a similar way?

Comment (flagarde) [reply cpp-terminal/platforms/exception.hpp:26]: @certik I fully agree with this and I tried to do it as much as possible. I try to find a way to avoid such things but this class is really windows specific. I could make it an useless class in Unix, but then it could be used by mistake in Unix specific code and could be har to debug. Please note that this is purely internal class and is only use by the developpers of cpp-terminal not the user (the .hpp is not installed) and it is not part of the public interface of the library. 

From a user perspective this class doesn't exist and he will catch the exception with the general Term::Exception. I designed it this way to be sure the class is not used in UNIX (code will not compile because the class would not exist) and to keep it internal to avoid user to use it in his own code.

The same apply to to_wide and to_narrow it is very specifics to Windows as the best approach up to know to have unicode is to use the *W functions that require `std::wstring` encoded in utf16. This function are purely internal too and trying to export then to UNIX would be useless and dangerous because wchar_t is 16bits in windows and 32bits in UNIX to my understanding. Puting then inside `#if defined(_WIN32)` assures this class is not available to UNIX world and as such gives a hint to the developper on the internal core library that he is certainly doing something wrong if he needs this function in UNIX.

If you really want to avoif the `#ifdef` in .h then the only possibility is to make them a dumb function in UNIX maybe with some `static_assert` or something similar.
